File: dvbViewerController/src/main/java/org/dvbviewer/controller/io/data/Status2Handler.java
/*
 * Copyright (C) 2012 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.io.data;

import android.sax.Element;
import android.sax.EndElementListener;
import android.sax.EndTextElementListener;
import android.sax.RootElement;
import android.sax.StartElementListener;
import android.util.Xml;

import org.dvbviewer.controller.R;
import org.dvbviewer.controller.entities.Status;
import org.dvbviewer.controller.entities.Status.Folder;
import org.dvbviewer.controller.entities.Status.StatusItem;
import org.xml.sax.Attributes;
import org.xml.sax.SAXException;
import org.xml.sax.helpers.DefaultHandler;

import java.util.ArrayList;

/**
 * The Class StatusHandler.
 *
 * @author RayBa
 * @date 01.07.2012
 */
public class Status2Handler extends DefaultHandler {

    Status status = null;
    Folder currentFolder = null;

    /**
     * Parses the.
     *
     * @param xml the xml
     * @return the status©
     * @throws SAXException
     * @author RayBa
     * @date 01.07.2012
     */
    public Status parse(String xml) throws SAXException {
        RootElement root = new RootElement("status");
        Element recordcount = root.getChild("reccount");
        Element streamclientcount = root.getChild("streamclientcount");
        Element rtspclientcount = root.getChild("rtspclientcount");
        Element unicastclientcount = root.getChild("unicastclientcount");
        Element lastuiaccess = root.getChild("lastuiaccess");
        Element nexttimer = root.getChild("nexttimer");
        Element nextrec = root.getChild("nextrec");
        Element standbyblock = root.getChild("standbyblock");
        Element tunercount = root.getChild("tunercount");
        Element streamtunercount = root.getChild("streamtunercount");
        Element rectunercount = root.getChild("rectunercount");
        Element recfiles = root.getChild("recfiles");
        Element epgudate = root.getChild("epgudate");
        Element recfolders = root.getChild("recfolders");
        Element folder = recfolders.getChild("folder");

        root.setStartElementListener(new StartElementListener() {

            @Override
            public void start(Attributes attributes) {
                status = new Status();
            }
        });

        recordcount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_current_recordings);
                item.setValue(body);
                status.getItems().add(item);
            }
        });

        streamclientcount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_current_clients);
                item.setValue(body);
                status.getItems().add(item);
            }

        });

        epgudate.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_epg_update_running);
                item.setValue(body);
                status.getItems().add(item);
            }

        });

        rtspclientcount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_current_rtsp_clients);
                item.setValue(body);
                status.getItems().add(item);
            }

        });

        unicastclientcount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_current_unicast_clients);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        nexttimer.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_next_timer);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        nextrec.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_next_Rec);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        lastuiaccess.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_last_ui_access);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        standbyblock.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_standby_blocked);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        tunercount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_tunercount);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        streamtunercount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_stream_tunercount);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        rectunercount.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_record_tunercount);
                item.setValue(body);
                status.getItems().add(item);
            }
        });
        recfiles.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                StatusItem item = new StatusItem();
                item.setNameRessource(R.string.status_recfiles);
                item.setValue(body);
                status.getItems().add(item);
            }
        });

        recfolders.setStartElementListener(new StartElementListener() {

            @Override
            public void start(Attributes attributes) {
                status.setFolders(new ArrayList<Folder>());
            }
        });
        folder.setStartElementListener(new StartElementListener() {

            @Override
            public void start(Attributes attributes) {
                currentFolder = new Folder();
                currentFolder.setSize(Long.valueOf(attributes.getValue("size")));
                currentFolder.setFree(Long.valueOf(attributes.getValue("free")));
            }
        });
        folder.setEndTextElementListener(new EndTextElementListener() {

            @Override
            public void end(String body) {
                currentFolder.setPath(body);
            }
        });
        folder.setEndElementListener(new EndElementListener() {

            @Override
            public void end() {
                status.getFolders().add(currentFolder);
            }
        });
        Xml.parse(xml, root.getContentHandler());
        return status;

    }

}

File: dvbViewerController/src/main/java/org/dvbviewer/controller/io/data/Status2Handler.kt
/*
 * Copyright (C) 2012 dvbviewer-controller Project
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package org.dvbviewer.controller.io.data

import android.sax.RootElement
import android.util.Xml
import org.apache.commons.lang3.math.NumberUtils
import org.dvbviewer.controller.R
import org.dvbviewer.controller.entities.Status
import org.dvbviewer.controller.entities.Status.Folder
import org.dvbviewer.controller.entities.Status.StatusItem
import org.xml.sax.SAXException
import org.xml.sax.helpers.DefaultHandler
import java.io.IOException
import java.io.InputStream
import java.util.*

/**
 * The Class Status1Handler.
 *
 * @author RayBa
 * @date 01.07.2012
 */
class Status2Handler : DefaultHandler(), StatusHandler {

    private var status: Status? = null
    private var currentFolder: Folder? = null

    /**
     * Parses the.
     *
     * @param stream the is
     * @return the status©
     * @throws SAXException
     * @author RayBa
     * @date 01.07.2012
     */
    @Throws(SAXException::class, IOException::class)
    override fun parse(stream: InputStream): Status? {
        val root = RootElement("status")
        val recordcount = root.getChild("reccount")
        val streamclientcount = root.getChild("streamclientcount")
        val rtspclientcount = root.getChild("rtspclientcount")
        val unicastclientcount = root.getChild("unicastclientcount")
        val lastuiaccess = root.getChild("lastuiaccess")
        val nexttimer = root.getChild("nexttimer")
        val nextrec = root.getChild("nextrec")
        val standbyblock = root.getChild("standbyblock")
        val tunercount = root.getChild("tunercount")
        val streamtunercount = root.getChild("streamtunercount")
        val rectunercount = root.getChild("rectunercount")
        val recfiles = root.getChild("recfiles")
        val epgudate = root.getChild("epgudate")
        val recfolders = root.getChild("recfolders")
        val folder = recfolders.getChild("folder")

        root.setStartElementListener { status = Status() }

        recordcount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_current_recordings
            item.value = body
            status!!.items.add(item)
        }

        streamclientcount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_current_clients
            item.value = body
            status!!.items.add(item)
        }

        epgudate.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_epg_update_running
            item.value = body
            status!!.items.add(item)
        }

        rtspclientcount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_current_rtsp_clients
            item.value = body
            status!!.items.add(item)
        }

        unicastclientcount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_current_unicast_clients
            item.value = body
            status!!.items.add(item)
        }
        nexttimer.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_next_timer
            item.value = body
            status!!.items.add(item)
        }
        nextrec.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_next_Rec
            item.value = body
            status!!.items.add(item)
        }
        lastuiaccess.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_last_ui_access
            item.value = body
            status!!.items.add(item)
        }
        standbyblock.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_standby_blocked
            item.value = body
            status!!.items.add(item)
        }
        tunercount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_tunercount
            item.value = body
            status!!.items.add(item)
        }
        streamtunercount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_stream_tunercount
            item.value = body
            status!!.items.add(item)
        }
        rectunercount.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_record_tunercount
            item.value = body
            status!!.items.add(item)
        }
        recfiles.setEndTextElementListener { body ->
            val item = StatusItem()
            item.nameRessource = R.string.status_recfiles
            item.value = body
            status!!.items.add(item)
        }

        recfolders.setStartElementListener { status!!.folders = ArrayList() }
        folder.setStartElementListener { attributes ->
            currentFolder = Folder()
            currentFolder!!.size = NumberUtils.toLong(attributes.getValue("size"))
            currentFolder!!.free = NumberUtils.toLong(attributes.getValue("free"))
        }
        folder.setEndTextElementListener { body -> currentFolder!!.path = body }
        folder.setEndElementListener { status!!.folders.add(currentFolder) }
        Xml.parse(stream, Xml.Encoding.UTF_8, root.contentHandler)
        return status

    }

}
